src/cargo/core/compiler/mod.rs

```diff
@@ -1177,6 +1251,56 @@ fn on_stderr_line(
     // Switch json lines from rustc/rustdoc that appear on stderr to stdout
     // instead. We want the stdout of Cargo to always be machine parseable as
     // stderr has our colorized human-readable messages.
-    state.stdout(&msg);
+    state.stdout(msg);
     Ok(())
 }
+
+fn replay_output_cache(
+    package_id: PackageId,
+    target: &Target,
+    path: PathBuf,
+    format: MessageFormat,
+    color: bool,
+) -> Work {
+    let target = target.clone();
+    let extract_rendered_messages = match format {
+        MessageFormat::Human => true,
+        MessageFormat::Json => false,
+        // FIXME: short not supported.
+        MessageFormat::Short => false,
+    };
+    Work::new(move |state| {
+        if path.exists() {
```

Comment:
I'm always a sucker for removing indentation, so this may be better as:

```rust
if !path.exists() {
    return Ok(()); // explanatory comment
}

// ...
```